src/instructlab/model/phased_training.py

```diff
@@ -0,0 +1,182 @@
+# Standard
+import datetime
+import enum
+import fcntl
+import logging
+import os
+import pathlib
+import typing
+import uuid
+
+# Third Party
+import pydantic
+import rich
+import yaml
+
+logger = logging.getLogger(__name__)
+
+
+class TrainingPhases(enum.Enum):
+    TRAIN1 = "train1"
+    TRAIN2 = "train2"
+    EVAL1 = "eval1"
+    EVAL2 = "eval2"
+    DONE = "done"
+
+
+AutoDatetimeField = pydantic.Field(
+    default_factory=lambda: datetime.datetime.now(datetime.UTC)
+)
+
+
+class EvalResult(pydantic.BaseModel):
+    """Single checkpoint's final score"""
+
+    ended_at_utc: datetime.datetime = AutoDatetimeField
+    checkpoint: pydantic.DirectoryPath
+    score: float
+
+    @pydantic.field_serializer("checkpoint", "ended_at_utc")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+
+class EvalPhaseModel(pydantic.BaseModel):
+    """Stores info about evaluation phase"""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: list[pydantic.DirectoryPath]
+    finished_checkpoints: list[pydantic.DirectoryPath] = []
+    results: list[EvalResult] = []
+    best_checkpoint: EvalResult | None = None
+
+    @pydantic.field_serializer(
+        "checkpoints",
+        "finished_checkpoints",
+    )
+    def pathlibPath_list_to_str(self, paths: list[pathlib.Path]) -> list[str]:
+        return [str(path) for path in paths]
+
+    @pydantic.field_serializer("ended_at_utc")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+
+class TrainPhaseModel(pydantic.BaseModel):
+    """Stores info about training phase. Doesn't store training args because user can get those elsewhere, e.g. config."""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: pydantic.DirectoryPath
+    # TODO: might want training args here, but can't currently do it
+    # without side-effects because model instance is changed inside of _training_phase
+
+    @pydantic.field_serializer("checkpoints", "ended_at_utc")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+
+class JournalModel(pydantic.BaseModel):
+    run_id: uuid.UUID = pydantic.Field(default_factory=uuid.uuid4)
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    current_phase: TrainingPhases = TrainingPhases.TRAIN1
+    train_1: TrainPhaseModel | None = None
+    eval_1: EvalPhaseModel | None = None
+    train_2: TrainPhaseModel | None = None
+    eval_2: EvalPhaseModel | None = None
+    final_output: EvalResult | None = None
+
+    @pydantic.field_serializer("current_phase")
+    def enum_to_value(self, enum_val: TrainingPhases) -> str:
+        return str(enum_val.value)
+
+
+class TrainingJournal:
```

Comment:
What type of "catastrophic failure" are you thinking about? I typically use a context manager when I need to ensure resources don't leak, such as when opening a file, managing a database connection, or acquiring a lock. In these scenarios, context managers are invaluable and we're already using them appropriately.

However, I think it's important to use context managers when their purpose is clearly visible (e.g., managing resources). Introducing them for more abstract scenarios might lead to unnecessary complexity, making the code harder to read or maintain. That said, if you believe the failure scenarios you're concerned about could benefit from the structured setup and teardown that a context manager provides, it might be worth considering. But simplicity and clarity should guide the decision.

Thanks!

Reply:
My personal preference is to not use a ctx manager for this. I like that committing the file to disk is an explicit action, plus the lifespan of an fp being open is only during the commit phase- transitively during write.

Reply:
In that case I think we should keep it as-is since that's simplest. We can revisit ways to optimize it if it becomes more painful in the future.